### ISSUE TITLE
SkyDNS doesn't retry to connect to etcd if it failed during startup

### ISSUE THREAD

**Issue by piosz** [GoogleCloudPlatform/kubernetes#5181]
It seems that SkyDNS doesn't try to reconnect to etcd if the client is broken (for example when etcd came up after SkyDNS).

There was a similar problem with kube2sky (#4814).

**Comment by thockin** [comment]
Rafal just fixed something like this - @rsokolowski , did you cut a new kube2sky?

I just realized you said skydns.  Is it skydns that dies or kube2sky?  

@bketelsen

**Comment by piosz** [comment]
The was a problem with SkyDNS. I was looking together with @rsokolowski and it seemed to be similar to the one with kube2sky that he fixed.

However as for now I can't reproduce the problem and everything looks good (I started skydns pod 100 times). So leaving this issue open for some time in case someone would have the same problem.

**Comment by piosz** [comment]
Successfully reproduced:
```
 $ kub log kube-dns-zx4b5 etcd
2015-03-10T19:05:29.740047256Z 2015/03/10 19:05:29 no data-dir provided, using default data-dir ./default.etcd
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcd: listening for peers on http://localhost:2380
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcd: listening for peers on http://localhost:7001
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcd: listening for client requests on http://0.0.0.0:2379
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcd: listening for client requests on http://0.0.0.0:4001
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: name = default
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: data dir = default.etcd
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: member dir = default.etcd/member
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: heartbeat = 100ms
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: election = 1000ms
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: snapshot count = 10000
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: advertise client URLs = http://127.0.0.1:4001
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2015-03-10T19:05:29.740134933Z 2015/03/10 19:05:29 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2015-03-10T19:05:29.759122285Z 2015/03/10 19:05:29 etcdserver: start member 6a5871dbdd12c17c in cluster f68652439e3f8f2a
2015-03-10T19:05:29.759122285Z 2015/03/10 19:05:29 raft: 6a5871dbdd12c17c became follower at term 0
2015-03-10T19:05:29.759122285Z 2015/03/10 19:05:29 raft: newRaft 6a5871dbdd12c17c [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015-03-10T19:05:29.759122285Z 2015/03/10 19:05:29 raft: 6a5871dbdd12c17c became follower at term 1
2015-03-10T19:05:29.769309796Z 2015/03/10 19:05:29 etcdserver: added local member 6a5871dbdd12c17c [http://localhost:2380 http://localhost:7001] to cluster f68652439e3f8f2a
2015-03-10T19:05:31.459036555Z 2015/03/10 19:05:31 raft: 6a5871dbdd12c17c is starting a new election at term 1
2015-03-10T19:05:31.459103080Z 2015/03/10 19:05:31 raft: 6a5871dbdd12c17c became candidate at term 2
2015-03-10T19:05:31.459125773Z 2015/03/10 19:05:31 raft: 6a5871dbdd12c17c received vote from 6a5871dbdd12c17c at term 2
2015-03-10T19:05:31.459171452Z 2015/03/10 19:05:31 raft: 6a5871dbdd12c17c became leader at term 2
2015-03-10T19:05:31.459196132Z 2015/03/10 19:05:31 raft.node: 6a5871dbdd12c17c elected leader 6a5871dbdd12c17c at term 2
2015-03-10T19:05:31.465936479Z 2015/03/10 19:05:31 etcdserver: published {Name:default ClientURLs:[http://127.0.0.1:4001]} to cluster f68652439e3f8f2a

$ kub log kube-dns-zx4b5 kube2sky
2015-03-10T19:05:48.666059731Z 2015/03/10 19:05:48 Etcd server found: http://127.0.0.1:4001
2015-03-10T19:05:48.666558522Z 2015/03/10 19:05:48 Using http://10.0.0.1:80 for kubernetes master
2015-03-10T19:05:48.666768428Z 2015/03/10 19:05:48 Using kubernetes API v1beta1
2015-03-10T19:05:48.691046005Z 2015/03/10 19:05:48 Setting dns record: elasticsearch-logging.default.kubernetes.local. -> 10.0.226.99:9200
2015-03-10T19:05:48.703084379Z 2015/03/10 19:05:48 Setting dns record: frontend.default.kubernetes.local. -> 10.0.254.137:8000
2015-03-10T19:05:48.717826518Z 2015/03/10 19:05:48 Setting dns record: kibana-logging.default.kubernetes.local. -> 10.0.142.67:5601
2015-03-10T19:05:48.733897825Z 2015/03/10 19:05:48 Setting dns record: kube-dns.default.kubernetes.local. -> 10.0.0.10:53
2015-03-10T19:05:48.749887980Z 2015/03/10 19:05:48 Setting dns record: kubernetes.default.kubernetes.local. -> 10.0.0.2:443
2015-03-10T19:05:48.774716453Z 2015/03/10 19:05:48 Setting dns record: kubernetes-ro.default.kubernetes.local. -> 10.0.0.1:80
2015-03-10T19:05:48.805852452Z 2015/03/10 19:05:48 Setting dns record: monitoring-grafana.default.kubernetes.local. -> 10.0.21.53:80
2015-03-10T19:05:48.822763323Z 2015/03/10 19:05:48 Setting dns record: monitoring-heapster.default.kubernetes.local. -> 10.0.0.112:80
2015-03-10T19:05:48.837760296Z 2015/03/10 19:05:48 Setting dns record: monitoring-influxdb.default.kubernetes.local. -> 10.0.252.64:80
2015-03-10T19:05:48.853706007Z 2015/03/10 19:05:48 Setting dns record: redis-master.default.kubernetes.local. -> 10.0.34.212:6379
2015-03-10T19:05:48.869983150Z 2015/03/10 19:05:48 Setting dns record: redis-slave.default.kubernetes.local. -> 10.0.9.49:6379

pszczesniak@piosz:~/git/kubernetes$ kub log kube-dns-zx4b5 skydns
2015-03-10T19:05:30.115981982Z [skydns] Mar 10 19:05:30.115 INFO      | falling back to default configuration, could not read from etcd:501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
2015-03-10T19:05:30.116227312Z [skydns] Mar 10 19:05:30.116 INFO      | ready for queries on kubernetes.local. for tcp://0.0.0.0:53 [rcache 0]
2015-03-10T19:05:30.116284069Z [skydns] Mar 10 19:05:30.116 INFO      | ready for queries on kubernetes.local. for udp://0.0.0.0:53 [rcache 0]
2015-03-10T19:10:10.265076392Z [skydns] Mar 10 19:10:10.264 INFO      | no nameservers defined or name too short, can not forward
2015-03-10T19:10:10.277797716Z [skydns] Mar 10 19:10:10.277 INFO      | no nameservers defined or name too short, can not forward
2015-03-10T19:10:14.373693976Z [skydns] Mar 10 19:10:14.373 INFO      | no nameservers defined or name too short, can not forward
2015-03-10T19:10:14.384716673Z [skydns] Mar 10 19:10:14.384 INFO      | no nameservers defined or name too short, can not forward
```

Can anyone confirm that kube-dns is working correctly (not just after start but generally)? Is it used by anything or as for now is it just running on kubernetes idly?

**Comment by Komnomnomnom** [comment]
Just saw this issue, didn't notice it when I created mine yesterday, d'oh. I was having similar problems and updating SkyDNS fixed it for me, see #5238 and skynetservices/skydns#142

**Comment by piosz** [comment]
Fixed by @Komnomnomnom in #5238. Thanks!
